File: android/src/main/java/com/red5pro/reactnative/view/R5VideoViewManager.java
package com.red5pro.reactnative.view;

import android.util.Log;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import com.red5pro.reactnative.module.R5StreamModule;
import com.red5pro.reactnative.stream.R5StreamInstance;
import com.red5pro.reactnative.stream.R5StreamMapManager;
import com.red5pro.reactnative.stream.R5StreamPublisher;
import com.red5pro.reactnative.stream.R5StreamSubscriber;
import com.red5pro.reactnative.util.RecordTypeUtil;
import com.red5pro.streaming.R5Stream;
import com.red5pro.streaming.R5StreamProtocol;
import com.red5pro.streaming.config.R5Configuration;

import java.util.HashMap;
import java.util.Map;

import javax.annotation.Nullable;

public class R5VideoViewManager extends SimpleViewManager<R5VideoViewLayout> implements
        R5StreamMapManager {

    private static final String TAG = "R5VideoViewManager";
    private static final String REACT_CLASS = "R5VideoView";

    private static final String PROP_HOST = "host";
    private static final String PROP_PORT = "port";
    private static final String PROP_CONTEXT_NAME = "contextName";
    private static final String PROP_STREAM_NAME = "streamName";
    private static final String PROP_BUFFER_TIME = "bufferTime";
    private static final String PROP_LICENSE_KEY = "licenseKey";
    private static final String PROP_BUNDLE_ID = "bundleID";
    private static final String PROP_PARAMETERS = "parameters";
    private static final String PROP_STREAM_BUFFER_TIME = "streamBufferTime";

    private static final int COMMAND_SUBSCRIBE = 1;
    private static final int COMMAND_PUBLISH = 2;
    private static final int COMMAND_UNSUBSCRIBE = 3;
    private static final int COMMAND_UNPUBLISH = 4;
    private static final int COMMAND_SWAP_CAMERA = 5;
    private static final int COMMAND_UPDATE_SCALE_MODE = 6;
    private static final int COMMAND_UPDATE_SCALE_SIZE = 7;
    private static final int COMMAND_MUTE_AUDIO = 8;
    private static final int COMMAND_UNMUTE_AUDIO = 9;
    private static final int COMMAND_MUTE_VIDEO = 10;
    private static final int COMMAND_UNMUTE_VIDEO = 11;
    private static final int COMMAND_SET_PLAYBACK_VOLUME = 12;
    private static final int COMMAND_DETACH = 13;
    private static final int COMMAND_ATTACH = 14;

    protected Map<String, R5StreamInstance> streamMap;

    public boolean addManagedStream (String name, R5StreamInstance item) {
        if (!streamMap.containsKey(name)) {
            streamMap.put(name, item);
            return true;
        }
        return false;
    }
    public boolean removeManagedStream (String name) {
        if (streamMap.containsKey(name)) {
            streamMap.remove(name);
            return true;
        }
        return false;
    }

    private R5VideoViewLayout mView;
    private ThemedReactContext mContext;
    private R5StreamModule mStreamModule;

    public void setModuleManager(R5StreamModule streamModule) {
        Log.d(TAG, "setModuleManager(): " + (streamModule == null));
        this.mStreamModule = streamModule;
    }

    public R5VideoViewManager(R5StreamModule streamModule) {
        super();
        Log.d(TAG, "got stream module: " + (streamModule == null));
        mStreamModule = streamModule;
        streamMap = new HashMap<>();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected R5VideoViewLayout createViewInstance(ThemedReactContext reactContext) {

        mContext = reactContext;
        mView = new R5VideoViewLayout(reactContext);
        return mView;

    }

    protected R5Configuration genereateConfiguration(ReadableMap configuration) {

        boolean hasHost = configuration.hasKey(PROP_HOST);
        boolean hasPort = configuration.hasKey(PROP_PORT);
        boolean hasContextName = configuration.hasKey(PROP_CONTEXT_NAME);
        boolean hasStreamName = configuration.hasKey(PROP_STREAM_NAME);
        boolean hasBufferTime = configuration.hasKey(PROP_BUFFER_TIME);
        boolean hasStreamBufferTime = configuration.hasKey(PROP_STREAM_BUFFER_TIME);
        boolean hasBundleID = configuration.hasKey(PROP_BUNDLE_ID);
        boolean hasLicenseKey = configuration.hasKey(PROP_LICENSE_KEY);
        boolean hasParameters = configuration.hasKey(PROP_PARAMETERS);

        boolean hasRequired = hasHost && hasPort && hasContextName;

        if (!hasRequired) {
            return null;
        }

        R5StreamProtocol protocol = R5StreamProtocol.RTSP;
        String host = configuration.getString(PROP_HOST);
        int port = configuration.getInt(PROP_PORT);
        String contextName = configuration.getString(PROP_CONTEXT_NAME);
        String streamName = configuration.getString(PROP_STREAM_NAME);
        String bundleID = hasBundleID ? configuration.getString(PROP_BUNDLE_ID) : "com.red5pro.android";
        String licenseKey = hasLicenseKey ? configuration.getString(PROP_LICENSE_KEY) : "";
        float bufferTime = hasBufferTime ? (float) configuration.getDouble(PROP_BUFFER_TIME) : 0.5f;
        float streamBufferTime = hasStreamBufferTime ? (float) configuration.getDouble(PROP_STREAM_BUFFER_TIME) : 2.0f;
        String parameters = hasParameters ? configuration.getString(PROP_PARAMETERS) : "";

        R5Configuration config = new R5Configuration(protocol, host, port, contextName, bufferTime, parameters);
        config.setStreamBufferTime(streamBufferTime);
        config.setBundleID(bundleID);
        config.setStreamName(streamName);
        config.setLicenseKey(licenseKey);

        return config;

    }

    @Override
    public void receiveCommand(final R5VideoViewLayout root, int commandId, @Nullable ReadableArray args) {

        Log.d(TAG, "Command(" + commandId + ")");
        if (args != null) {
            Log.d(TAG, "Args are " + args.toString());
        }

        switch (commandId) {
            case COMMAND_UPDATE_SCALE_SIZE:

                int updateWidth = args.getInt(0);
                int updateHeight = args.getInt(1);
                int screenWidth = args.getInt(2);
                int screenHeight = args.getInt(3);
                root.updateScaleSize(updateWidth, updateHeight, screenWidth, screenHeight);

                break;
            case COMMAND_SUBSCRIBE:

                final String streamName = args.size() > 0
                        ? args.getString(0) : root.getConfiguration().getStreamName();

                R5StreamInstance instance;
                if (!streamMap.containsKey(streamName)) {
                    instance = new R5StreamSubscriber(mContext);
                    streamMap.put(streamName, instance);
                    root.setStreamInstance(instance);
                    root.subscribe(streamName);
                } else {
                    instance = streamMap.get(streamName);
                    root.setStreamInstance(instance);
                    if (root.mAttached) {
                        root.attach();
                    }
                }

                break;
            case COMMAND_UNSUBSCRIBE:

                final String unsubscribe_streamName = args.size() > 0
                        ? args.getString(0) : root.getConfiguration().getStreamName();

                R5StreamInstance unsubscribe_instance;
                if (streamMap.containsKey(unsubscribe_streamName)) {
                    unsubscribe_instance = streamMap.get(unsubscribe_streamName);
                    ((R5StreamSubscriber)unsubscribe_instance).unsubscribe();
                    root.unsubscribe();
                    root.setStreamInstance(null);
                    streamMap.remove(unsubscribe_streamName);
                }

                break;
            case COMMAND_PUBLISH:

                final int type = args.getInt(1);

                R5Stream.RecordType recordType = RecordTypeUtil.typeFromJSEnumValue(type);

                final String pub_streamName = args.size() > 0
                        ? args.getString(0) : root.getConfiguration().getStreamName();

                R5StreamInstance pub_instance;
                if (!streamMap.containsKey(pub_streamName)) {
                    pub_instance = new R5StreamPublisher(mContext);
                    streamMap.put(pub_streamName, pub_instance);
                    root.setStreamInstance(pub_instance);
                    root.publish(pub_streamName, recordType);
                } else {
                    pub_instance = streamMap.get(pub_streamName);
                    root.setStreamInstance(pub_instance);
                    if (root.mAttached) {
                        root.attach();
                    }
                }

                break;
            case COMMAND_UNPUBLISH:

                final String unpublish_streamName = args.size() > 0
                        ? args.getString(0) : root.getConfiguration().getStreamName();

                R5StreamInstance unpublish_instance;
                if (streamMap.containsKey(unpublish_streamName)) {
                    unpublish_instance = streamMap.get(unpublish_streamName);
                    ((R5StreamPublisher)unpublish_instance).unpublish();
                    root.unpublish();
                    root.setStreamInstance(null);
                    streamMap.remove(unpublish_streamName);
                }

                break;
            case COMMAND_SWAP_CAMERA:

                root.swapCamera();

                break;
            case COMMAND_UPDATE_SCALE_MODE:

                final int mode = args.getInt(0);
                root.updateScaleMode(mode);

                break;
            case COMMAND_MUTE_AUDIO:

                root.muteAudio();

                break;
            case COMMAND_UNMUTE_AUDIO:

                root.unmuteAudio();

                break;
            case COMMAND_MUTE_VIDEO:

                root.muteVideo();

                break;
            case COMMAND_UNMUTE_VIDEO:

                root.unmuteVideo();

                break;
            case COMMAND_SET_PLAYBACK_VOLUME:

                final int value = args.getInt(0);
                root.setPlaybackVolume(value/100);

                break;
            case COMMAND_DETACH:

                Log.d(TAG, ":detach()");

                root.detach();

                final String detach_stream_id = args.getString(0);
                R5StreamInstance detach_stream = mStreamModule.getStreamInstance(detach_stream_id);
                if (detach_stream != null) {
                    root.setStreamInstance(null);
                } else {
                    Log.d(TAG, "Could not detach. No matching stream with id:(" + detach_stream_id + ") stored.");
                }

                break;
            case COMMAND_ATTACH:

                Log.d(TAG, ":attach()");

                final String attach_stream_id = args.getString(0);
                R5StreamInstance attach_stream = mStreamModule.getStreamInstance(attach_stream_id);
                if (attach_stream != null) {
                    root.setStreamInstance(attach_stream);
                    root.attach();
                } else {
                    Log.d(TAG, "Could not attach. No matching stream with id:(" + attach_stream_id + ") stored.");
                }

                break;
            default:
                super.receiveCommand(root, commandId, args);
                break;
        }
    }

    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        MapBuilder.Builder<String, Integer> builder = MapBuilder.builder();
        for (R5VideoViewLayout.Commands command : R5VideoViewLayout.Commands.values()) {
            builder.put(command.toString(), command.getValue());
        }
        return builder.build();
    }

    @Override
    @Nullable
    public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
        MapBuilder.Builder<String, Object> builder = MapBuilder.builder();
        for (R5VideoViewLayout.Events event : R5VideoViewLayout.Events.values()) {
            builder.put(event.toString(), MapBuilder.of("registrationName", event.toString()));
        }
        return builder.build();
    }

    private R5Configuration createConfigurationFromMap(ReadableMap configuration) {

        boolean hasHost = configuration.hasKey(PROP_HOST);
        boolean hasPort = configuration.hasKey(PROP_PORT);
        boolean hasContextName = configuration.hasKey(PROP_CONTEXT_NAME);
        boolean hasStreamName = configuration.hasKey(PROP_STREAM_NAME);
        boolean hasBufferTime = configuration.hasKey(PROP_BUFFER_TIME);
        boolean hasStreamBufferTime = configuration.hasKey(PROP_STREAM_BUFFER_TIME);
        boolean hasBundleID = configuration.hasKey(PROP_BUNDLE_ID);
        boolean hasLicenseKey = configuration.hasKey(PROP_LICENSE_KEY);
        boolean hasParameters = configuration.hasKey(PROP_PARAMETERS);

        boolean hasRequired = hasHost && hasPort && hasContextName;

        if (!hasRequired) {
            return null;
        }

        R5StreamProtocol protocol = R5StreamProtocol.RTSP;
        String host = configuration.getString(PROP_HOST);
        int port = configuration.getInt(PROP_PORT);
        String contextName = configuration.getString(PROP_CONTEXT_NAME);
        String streamName = hasStreamName ? configuration.getString(PROP_STREAM_NAME) : "mystream";
        String bundleID = hasBundleID ? configuration.getString(PROP_BUNDLE_ID) : "com.red5pro.android";
        String licenseKey = hasLicenseKey ? configuration.getString(PROP_LICENSE_KEY) : "";
        float bufferTime = hasBufferTime ? (float) configuration.getDouble(PROP_BUFFER_TIME) : 1.0f;
        float streamBufferTime = hasStreamBufferTime ? (float) configuration.getDouble(PROP_STREAM_BUFFER_TIME) : 2.0f;
        String parameters = hasParameters ? configuration.getString(PROP_PARAMETERS) : "";

        Log.d(TAG, "Parameters: " + parameters);

        R5Configuration config = new R5Configuration(protocol, host, port, contextName, bufferTime, parameters);

        config.setStreamBufferTime(streamBufferTime);
        config.setBundleID(bundleID);
        config.setStreamName(streamName);
        config.setLicenseKey(licenseKey);

        return config;

    }

    @ReactProp(name = "configuration")
    public void setConfiguration(R5VideoViewLayout view, ReadableMap configuration) {
        R5Configuration conf = createConfigurationFromMap(configuration);
        boolean autoAttachView = configuration.hasKey("autoAttachView") ? configuration.getBoolean("autoAttachView") : true;
        view.updateConfiguration(conf, configuration.getString("key"), autoAttachView);
    }

    @ReactProp(name = "showDebugView", defaultBoolean = false)
    public void setShowDebugView(R5VideoViewLayout view, boolean showDebug) {
        view.updateShowDebug(showDebug);
    }

    @ReactProp(name = "scaleMode", defaultInt = 0) // 0, 1, 2
    public void setScaleMode(R5VideoViewLayout view, int mode) {
        view.updateScaleMode(mode);
    }

    @ReactProp(name = "logLevel", defaultInt = 3) // LOG_LEVEL_ERROR
    public void setLogLevel(R5VideoViewLayout view, int logLevel) {
        view.updateLogLevel(logLevel);
    }

    @ReactProp(name = "publishVideo", defaultBoolean = true)
    public  void setPublishVideo(R5VideoViewLayout view, boolean useVideo) {
        view.updatePublishVideo(useVideo);
    }

    @ReactProp(name = "publishAudio", defaultBoolean = true)
    public  void setPublishAudio(R5VideoViewLayout view, boolean useAudio) {
        view.updatePublishAudio(useAudio);
    }

    @ReactProp(name = "subscribeVideo", defaultBoolean = true)
    public  void setSubscribeVideo(R5VideoViewLayout view, boolean playbackVideo) {
        view.updateSubscribeVideo(playbackVideo);
    }

    @ReactProp(name = "cameraWidth", defaultInt = 640)
    public void setCameraWidth(R5VideoViewLayout view, int value) {
        view.updateCameraWidth(value);
    }

    @ReactProp(name = "cameraHeight", defaultInt = 360)
    public void setCameraHeight(R5VideoViewLayout view, int value) {
        view.updateCameraHeight(value);
    }

    @ReactProp(name = "bitrate", defaultInt = 750)
    public void setBitrate(R5VideoViewLayout view, int value) {
        view.updatePublishBitrate(value);
    }

    @ReactProp(name = "framerate", defaultInt = 15)
    public void setFramerate(R5VideoViewLayout view, int value) {
        view.updatePublishFramerate(value);
    }

    @ReactProp(name = "audioBitrate", defaultInt = 32)
    public void setAudioBitrate(R5VideoViewLayout view, int value) {
        view.updatePublishAudioBitrate(value);
    }

    @ReactProp(name = "audioSampleRate", defaultInt = 44100)
    public void setAudioSampleRate(R5VideoViewLayout view, int value) {
        view.updatePublishAudioSampleRate(value);
    }

    /*
     *
        public static enum PlaybackMode {
            AEC,
            STANDARD;

            private PlaybackMode() {}
        }
     *
     */
    @ReactProp(name = "audioMode", defaultInt = 0)
    public void setSubscriberAudioMode(R5VideoViewLayout view, int value) {
        view.updateSubscriberAudioMode(value);
    }

    @ReactProp(name = "useAdaptiveBitrateController", defaultBoolean = false)
    public void setUseAdaptiveBitrateController(R5VideoViewLayout view, boolean value) {
        view.updatePublisherUseAdaptiveBitrateController(value);
    }

    @ReactProp(name = "useBackfacingCamera", defaultBoolean = false)
    public void setUseBackfacingCamera(R5VideoViewLayout view, boolean value) {
        view.updatePublisherUseBackfacingCamera(value);
    }

    @ReactProp(name = "enableBackgroundStreaming", defaultBoolean = false)
    public void setEnableBackgroundStreaming(R5VideoViewLayout view, boolean value) {
        view.updatePubSubBackgroundStreaming(value);
    }

    @ReactProp(name = "zOrderOnTop", defaultBoolean = true)
    public  void setZOrderOnTop(R5VideoViewLayout view, boolean value) {
        view.updateZOrderOnTop(value);
    }

    @ReactProp(name = "zOrderMediaOverlay", defaultBoolean = true)
    public  void setZOrderOverlayMedia(R5VideoViewLayout view, boolean value) {
        view.updateZOrderMediaOverlay(value);
    }

    @Nullable
    @Override
    public Map<String, Object> getConstants() {
        return super.getConstants();
    }

    @Override
    public boolean hasConstants() {
        return super.hasConstants();
    }

}

File: ios/R5VideoView/R5VideoView/R5VideoViewManager.m
//
//  R5VideoViewManager.m
//  React Native Red5Pro
//
//  Created by Todd Anderson on 10/27/17.
//  Copyright © 2017 Infrared5, Inc. All rights reserved.
//

#import <React/RCTViewManager.h>
#import <R5Streaming/R5Streaming.h>
#import <React/RCTUIManager.h>

#import "R5VideoView.h"
#import "R5StreamItem.h"
#import "R5StreamModule.h"
#import "R5StreamPublisher.h"
#import "R5StreamSubscriber.h"
#import "R5VideoViewManager.h"

static NSMutableDictionary *_streamMap;

@implementation R5VideoViewManager

RCT_EXPORT_MODULE()

// TODO:
//+ (BOOL)requiresMainQueueSetup
//{
//    return NO;
//}

# pragma RN Events
RCT_EXPORT_VIEW_PROPERTY(onConfigured, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onMetaDataEvent, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPublisherStreamStatus, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onSubscriberStreamStatus, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUnpublishNotification, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onUnsubscribeNotification, RCTBubblingEventBlock)

RCT_EXPORT_METHOD(subscribe:(nonnull NSNumber *)reactTag streamName:(nonnull NSString *)streamName) {
    
    RCTLogInfo(@"R5VideoViewManager:subscribe(%@)", streamName);
    NSMutableDictionary *map = [R5StreamModule streamMap];
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            R5StreamItem *item = [map objectForKey:streamName];
            if (item == nil || [item getStreamInstance] == nil) {
                RCTLog(@"Creating new subscriber instance for %@", streamName);
                R5StreamItem *newItem = [[R5StreamItem alloc] initWithConfiguration:[view configuration]];
                R5StreamSubscriber *subscriber = [[R5StreamSubscriber alloc] init];
                [newItem setStreamInstance:subscriber];
                [map setObject:newItem forKey:streamName];
                [view setStreamInstance:(NSObject<R5StreamInstance> *)subscriber];
                [view subscribe:streamName];
            } else {
                [view setStreamInstance:[item getStreamInstance]];
                if ([view getIsAttached]) {
                    [view attach];
                }
            }
        }
    }];
    
}

RCT_EXPORT_METHOD(unsubscribe:(nonnull NSNumber *)reactTag) {
    
    RCTLogInfo(@"R5VideoViewManager:unsubscribe()");
    NSMutableDictionary *map = [R5StreamModule streamMap];
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            NSString *streamName = [view.configuration streamName];
            R5StreamItem *item = [map objectForKey:streamName];
            if (item != nil) {
                [view unsubscribe];
                [item clear];
                [map removeObjectForKey:streamName];
            } else {
                [view unsubscribe];
            }
        }
    }];
    
}

RCT_EXPORT_METHOD(publish:(nonnull NSNumber *)reactTag streamName:(nonnull NSString *)streamName withMode:(int)publishMode) {
    
    RCTLogInfo(@"R5VideoViewManager:publish(%@)", streamName);
    NSMutableDictionary *map = [R5StreamModule streamMap];
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            R5StreamItem *item = [map objectForKey:streamName];
            if (item == nil || [item getStreamInstance] == nil) {
                RCTLog(@"Creating new publisher instance for %@", streamName);
                R5StreamItem *newItem = [[R5StreamItem alloc] initWithConfiguration:[view configuration]];
                R5StreamPublisher *publisher = [[R5StreamPublisher alloc] init];
                [item setStreamInstance:publisher];
                [map setObject:newItem forKey:streamName];
                [view setStreamInstance:(NSObject<R5StreamInstance> *)publisher];
                [view publish:streamName withMode:publishMode];
            } else {
                NSObject<R5StreamInstance> *stream = [item getStreamInstance];
                [view setStreamInstance:stream];
                if ([view getIsAttached]) {
                    [view attach];
                }
            }
        }
    }];
    
}

RCT_EXPORT_METHOD(unpublish:(nonnull NSNumber *)reactTag) {
    
    RCTLogInfo(@"R5VideoViewManager:unpublish()");
    NSMutableDictionary *map = [R5StreamModule streamMap];
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            R5Configuration *configuration = [view configuration];
            NSString *streamName = [configuration streamName];
            R5StreamItem *item = [map objectForKey:streamName];
            if (item != nil) {
//                NSObject<R5StreamInstance> *stream = [item getStreamInstance];
//                if (stream != nil) {
//                    [(R5StreamPublisher *)stream unpublish];
//                    [stream setEmitter:nil];
//                }
                [view unpublish];
                [item clear];
//                [item setStreamInstance:nil];
                [map removeObjectForKey:streamName];
            } else {
                [view unpublish];
            }
        }
    }];
    
}

RCT_EXPORT_METHOD(swapCamera:(nonnull NSNumber *)reactTag) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view swapCamera];
        }
    }];
    
}

RCT_EXPORT_METHOD(updateScaleMode:(nonnull NSNumber *)reactTag mode:(int)mode) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view updateScaleMode:mode];
        }
    }];
    
}

RCT_EXPORT_METHOD(updateScaleSize:(nonnull NSNumber *)reactTag width:(int)width withHeight:(int)height withScreenWidth:(int)screenWidth withScreenHeight:(int)screenHeight) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view updateScaleSize:width withHeight:height withScreenWidth:screenWidth withScreenHeight:screenHeight];
        }
    }];
    
}

RCT_EXPORT_METHOD(muteAudio:(nonnull NSNumber *)reactTag) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view muteAudio];
        }
    }];
    
}

RCT_EXPORT_METHOD(unmuteAudio:(nonnull NSNumber *)reactTag) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view unmuteAudio];
        }
    }];
    
}

RCT_EXPORT_METHOD(muteVideo:(nonnull NSNumber *)reactTag) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view muteVideo];
        }
    }];
    
}

RCT_EXPORT_METHOD(unmuteVideo:(nonnull NSNumber *)reactTag) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view unmuteVideo];
        }
    }];
    
}

RCT_EXPORT_METHOD(setPlaybackVolume:(nonnull NSNumber *)reactTag value:(int)value) {
    
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
        R5VideoView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[R5VideoView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
        } else {
            [view setPlaybackVolume:value];
        }
    }];
    
}

RCT_EXPORT_METHOD(attach:(nonnull NSNumber *)reactTag withId:(NSString* )streamId) {
    NSMutableDictionary *map = [R5StreamModule streamMap];
    if ([map objectForKey:streamId] != nil) {
        R5StreamItem *item = [map objectForKey:streamId];
        [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
            R5VideoView *view = viewRegistry[reactTag];
            if (![view isKindOfClass:[R5VideoView class]]) {
                RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
            } else {
                RCTLog(@"Found view for instance attach(%@).", streamId);
                [view setStreamInstance:[item getStreamInstance]];
                [view attach];
            }
        }];
    }
}

RCT_EXPORT_METHOD(detach:(nonnull NSNumber *)reactTag withId:(NSString* )streamId) {
    NSMutableDictionary *map = [R5StreamModule streamMap];
    if ([map objectForKey:streamId] != nil) {
        [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, R5VideoView *> *viewRegistry) {
            R5VideoView *view = viewRegistry[reactTag];
            if (![view isKindOfClass:[R5VideoView class]]) {
                RCTLogError(@"Invalid view returned from registry, expecting R5VideoView, got: %@", view);
            } else {
                RCTLog(@"Found view for instance detach(%@).", streamId);
                [view detach];
            }
        }];
    }
}

# pragma RN Properties
RCT_EXPORT_VIEW_PROPERTY(logLevel, int);
RCT_EXPORT_VIEW_PROPERTY(scaleMode, int);
RCT_EXPORT_VIEW_PROPERTY(publishVideo, BOOL);
RCT_EXPORT_VIEW_PROPERTY(publishAudio, BOOL);
RCT_EXPORT_VIEW_PROPERTY(subscribeVideo, BOOL);
RCT_EXPORT_VIEW_PROPERTY(cameraWidth, int);
RCT_EXPORT_VIEW_PROPERTY(cameraHeight, int);
RCT_EXPORT_VIEW_PROPERTY(bitrate, int);
RCT_EXPORT_VIEW_PROPERTY(framerate, int);
RCT_EXPORT_VIEW_PROPERTY(audioMode, int);
RCT_EXPORT_VIEW_PROPERTY(audioBitrate, int);
RCT_EXPORT_VIEW_PROPERTY(audioSampleRate, int);
RCT_EXPORT_VIEW_PROPERTY(useBackfacingCamera, BOOL);
RCT_EXPORT_VIEW_PROPERTY(useAdaptiveBitrateController, BOOL);
RCT_EXPORT_VIEW_PROPERTY(enableBackgroundStreaming, BOOL);

RCT_CUSTOM_VIEW_PROPERTY(showDebugView, BOOL, R5VideoView) {
    [view setShowDebugInfo:[json boolValue]];
}

RCT_CUSTOM_VIEW_PROPERTY(configuration, R5Configuration, R5VideoView) {
  
    R5Configuration *configuration = [[R5Configuration alloc] init];
    configuration.protocol = 1;
    configuration.host = json[@"host"];
    configuration.port = [json[@"port"] intValue];
    configuration.contextName = json[@"contextName"];
    configuration.streamName = json[@"streamName"];
    configuration.bundleID = json[@"bundleID"];
    configuration.licenseKey = json[@"licenseKey"];
    configuration.buffer_time = [json[@"bufferTime"] floatValue];
    configuration.stream_buffer_time = [json[@"streamBufferTime"] floatValue];
    configuration.parameters = json[@"parameters"];
  
    BOOL autoAttach = YES;
    if (json[@"autoAttachView"] != nil) {
        autoAttach = [json[@"autoAttachView"] boolValue];
    }
    [view loadConfiguration:configuration forKey:json[@"key"] andAttach:autoAttach];

}

- (UIView *)view {
  R5VideoView *r5View = [[R5VideoView alloc] init];
  return r5View;
}

@end
